### DISCUSSION TITLE
Deploy without Azure Firewall and Private DNS Zones?

### DISCUSSION THREAD

**Discussion by baatch** [Azure/data-management-zone#211] [Q&A] [answered]
In the guidance docs for cost estimates (https://github.com/Azure/data-management-zone/blob/main/docs/guidance/EnterpriseScaleAnalytics-CostEstimates.md) there are references to:

Pricing Calculator - Data Management Zone w/o Azure Firewall and Private DNS Zones
Pricing Calculator - Data Management Zone w/ Azure Firewall and Private DNS Zones

How do you deploy the solution without Azure Firewall and Private DNS Zones?

In the params.dev.json file I could find the parameters "enableDnsAndFirewallDeployment" but even when setting that to false I still get deployment errors.

![image](https://user-images.githubusercontent.com/16446781/139246820-26e98b33-8bb9-464e-951c-10040647140c.png)

`"message": "{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"DeploymentFailed\",\r\n        \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.\",\r\n        \"details\": [\r\n          {\r\n            \"code\": \"BadRequest\",\r\n            \"message\": \"{\\r\\n  \\\"error\\\": {\\r\\n    \\\"code\\\": \\\"LinkedInvalidPropertyId\\\",\\r\\n    \\\"message\\\": \\\"Property id '' at path 'properties.privateDnsZoneConfigs[0].properties.privateDnsZoneId' is invalid. Expect fully qualified resource Id that start with '/subscriptions/{subscriptionId}' or '/providers/{resourceProviderNamespace}/'.\\\"\\r\\n  }\\r\\n}\"\r\n          }\r\n        ]\r\n      }\r\n    ]\r\n  }\r\n}"`

Some documentation around how to setup this would be helpful as deploying a POC/MVP that cost 1400$ per month is a no go for many companies.

**Comment by marvinbuss** [comment, marked as answer]
Hi @baatch,
Can you please click on `containerResources` and then on `containerRegistry001` and provide me the error message, so that I can fix the issue?
Feel free to also ping me on Teams, if you are a Microsoft employee, to troubleshoot this quickly.
Thanks for your help!

**Reply by baatch** [reply]
Hi @marvinbuss 

Sorry not MS employee but can connect on Teams/LinkedIn.

The error message for containerResources was in initial message: 

`"message": "{\r\n \"status\": \"Failed\",\r\n \"error\": {\r\n \"code\": \"ResourceDeploymentFailure\",\r\n \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n \"details\": [\r\n {\r\n \"code\": \"DeploymentFailed\",\r\n \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.\",\r\n \"details\": [\r\n {\r\n \"code\": \"BadRequest\",\r\n \"message\": \"{\\r\\n \\\"error\\\": {\\r\\n \\\"code\\\": \\\"LinkedInvalidPropertyId\\\",\\r\\n \\\"message\\\": \\\"Property id '' at path 'properties.privateDnsZoneConfigs[0].properties.privateDnsZoneId' is invalid. Expect fully qualified resource Id that start with '/subscriptions/{subscriptionId}' or '/providers/{resourceProviderNamespace}/'.\\\"\\r\\n }\\r\\n}\"\r\n }\r\n ]\r\n }\r\n ]\r\n }\r\n}"`

**Reply by marvinbuss** [reply]
Thanks @baatch! I already found the issue. Sadly, there is a small bug in the container deployment. Will open a PR and fix that quickly.

Did you also receive any additional error message in the `governanceServices` deployments? Would like to fix both of the quickly.
I am very sorry for the inconveniences caused.

**Reply by marvinbuss** [reply]
Feel free to ping me via LinkedIn or Teams. 👍

**Reply by marvinbuss** [reply]
You can track progress here: https://github.com/Azure/data-management-zone/pull/215
We still need to investigate the Purview Issue.

**Reply by baatch** [reply]
@marvinbuss 

Here is the issue with the governanceServices:

`{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"InvalidTemplateDeployment","message":"The template deployment failed with error: 'Authorization failed for template resource '9753a46b-5755-5850-bd89-0bd35f0ab173' of type 'Microsoft.Authorization/roleAssignments'. The client '1733a64c-81f9-4d44-a2a9-4b342e7e587f' with object id '1733a64c-81f9-4d44-a2a9-4b342e7e587f' does not have permission to perform action 'Microsoft.Authorization/roleAssignments/write' at scope '/subscriptions/48fc3def-3fff-4933-afcf-cdcc2a8da06a/resourceGroups/kexdmz-dev-governance/providers/Microsoft.KeyVault/vaults/kexdmz-dev-vault001/providers/Microsoft.Authorization/roleAssignments/9753a46b-5755-5850-bd89-0bd35f0ab173'.'."}]}`


![image](https://user-images.githubusercontent.com/16446781/139274223-efa150ee-28da-49bf-9c20-f55a082e57b0.png)

**Reply by marvinbuss** [reply]
@baatch This seems o be a permissions error in your subscription. It looks like you do not have the rights to make role assignments.

**Reply by marvinbuss** [reply]
@baatch I have thought about this and I do not necessarily want to make role assignments optional in the Data Landing Zone or Data Management Zone as this would lead to a half-bakes setup. Could you ask your administrator to grant you either `Owner` or `User Access Administrator` in the subscription to which you are trying to deploy the setup?

**Reply by marvinbuss** [reply]
https://github.com/Azure/data-management-zone/pull/215 is now merged. Therefore, you should not see the container registry issue from now on. Sorry for the bug. Please let me know, whether you can get additional access rights in your subscription.

**Reply by baatch** [reply]
Great @marvinbuss! 

I'm redeployed and there are no issues now. I missed the User Access Administrator assignment before. Thanks for the quick fix =)  

I'm not really a network person but with this option set as false is it correctly understood that everything will still work with vnets and private endpoints but just no Global DNS and Azure Firewall?

And is the assumption that I can just deploy the Data Landing Zone and connect it to the Data Management Zone I just deployed and it will work as expected?

**Reply by marvinbuss** [reply]
Thanks for confirming! 👍 
I am glad that it all works now.

The Azure Firewall is required for any outbound traffic from the respective services (Data Factory, Azure Databricks, Azure Machine Learning, etc.). You can either rely on an existing Firewall in the tenant (most customers have a central NVA already deployed) or you deploy it as part of ESA. If the central Firewall is not available, outbound traffic may not work as expected.

Private DNS Zones are required for resolving the FQDNs to the private IPs of the private endpoints. Again, if customer already have an existing DNS infrastructure, ESA allows to make use of it. If none of this is available the DNS resolution to the private IPs of the private endpoints will not work and hence connectivity to services like storage or Synapse or other services will fail as they are all blocking public network access.

In our case, we are using the Azure Firewall also as DNS forwarder. If there is no DNS forwarder available, there will be no option to centrally resolve all DNS requests from the Data Landing Zones. Again, ESA allows to make use of existing infrastructure. If nothing of this is available, DNS resolution will again not work.

Some Firewall, DNS infrastructure and DNS forwarder is required. Eitehr it gets deployed through ESA or exusting infrastructure is used within the tenant. I hope that is clear?

**Reply by baatch** [reply]
@marvinbuss  thanks, I think it is a little more clear. 

But still my understanding is that there needs to deploy/configure something in place of the Azure Firewall and Global DNS if they are not included in the deployment to be able to fully evaluate the solution correct? 

Can you build or recommend alternative solutions (linux firewalls) that will fit into the ESA and can be used for this scenario (small POC/MVP) that will be cheaper than 1400$ /month? 

Otherwise I would think this can be a huge barrier for many to both adopt and learn if you need to get a budget of 1400$ / month just for the data management zone.

**Reply by marvinbuss** [reply]
Yes, that is correct. If you are not usng our Firewall, then you need to use an alternative.
In most environments and tenants, there is already an existing shared network infrastructure that can be used (Firewall, DNS, etc.).

If we move to a different solution, user will have to start manage the solution. Also, there is most likely no native ARM integration that we will be able to use to provide the same one-click experience. If you want to save some money, you can deploy ESA via CI/CD and delete the Firewall at the end of the day and then redeploy, when you start working in the morning through a pipeline.
Would that work?

I do understand that for a small use case team that just looks at integrating a dataset and developing some workflows, this may look expensive. However, we are targeting the data platform layer, which may span hundreds or thousands of use cases over time. For those scenarios, this should not be the biggest blocker.

Also, I believe that we should not trade less cost for much higher security risks.. We should be more concerned about cost and proper security measures and less about cost in these scenarios.

**Reply by baatch** [reply]
**_If you want to save some money, you can deploy ESA via CI/CD and delete the Firewall at the end of the day and then redeploy, when you start working in the morning through a pipeline.
Would that work?_**

By doing this, you mentioned that outbound traffic may not work? Can the solution still be evaluated fully and properly?

I agree that you should not trade cost for security and I fully understand that this is not targeting small teams/companies. 

The larger enterprise companies and customers I'm working with are fine with the cost of the Azure firewall in production, but before getting that far they usually want try/evaluate the solution/architecture first to prove it's value. Can you provide guidance, tips or develop something for these scenarios?

**Reply by marvinbuss** [reply]
That is correct. Running workloads while the Azure Firewall is deleted will also fail, as the DNS resolution will not work. Therefore, the suggestion would only be an option while developing.
Thank you for acknowledging the trade-off  between cost and security. 👍 

What would you like to see from a value standpoint?
I believe that the current cost is justifiable if you are developing a new data platform for your organization or business.
